pkg/sentry/platform/ring0/defs_amd64.go

```diff
@@ -72,19 +72,59 @@ type KernelOpts struct {
 	PageTables *pagetables.PageTables
 }
 
+// Currently, KVM_MAX_VCPUS is 288, which has been being increased slowly.
+// So 512 for _RING0_MAX_CPUS is enough for a while.
+const _RING0_MAX_CPUS = 512
+
+type cpuEntryArea struct {
```

Comment:
Can we just ensure that the full kvm.CPU is also mapped in the upper half? This will include the kernel stack and all the other bits that I think matter, and we won't need to introduce an additional structures.

Though I believe that we will also need to load registers before switching CR3s, but that can be done.

Reply:
(I think you're pushing the kernelCR3 down into the ring0.CPU struct and making that switch part of the entry/exit routines, which is a fine solution and eliminates the register problem anyways.)

Reply:
.... And you're ahead of me. You've done it already below.

Reply:
ring0.CPU is still a big that contains sensitive info (the whole sentry kernel registers etc), which is one of the reason I split kernelEntry out from ring0.CPU. Other reason is that I don't want to change the allocation of kvm.vCPU which is none-arch independence code; the structure mapped on the upper should use it's own allocation, although it is a simple one by now. I'm planing to make idt gdt tss readonly in upper, kernelEntry will be much larger and we need to really implement the allocation.

Reply:
> Why the test fails?
> How do other PRs survive this this test:
> 
> > Use --sandbox_debug to see verbose messages from the sandbox
> > compilepkg: nogo: errors found by nogo during build-time code analysis:
> > /home/travis/.cache/bazel/_bazel_gvisor/bcc3876c18b4288f70c2594c9439fc99/sandbox/processwrapper-sandbox/5/execroot/**main**/pkg/sentry/platform/kvm/machine_arm64_unsafe.go:52:29: possible misuse of unsafe.Pointer
> > /home/travis/.cache/bazel/_bazel_gvisor/bcc3876c18b4288f70c2594c9439fc99/sandbox/processwrapper-sandbox/5/execroot/**main**/pkg/sentry/platform/kvm/machine_arm64_unsafe.go:90:21: possible misuse of unsafe.Pointer
> > /home/travis/.cache/bazel/_bazel_gvisor/bcc3876c18b4288f70c2594c9439fc99/sandbox/processwrapper-sandbox/5/execroot/**main**/pkg/sentry/platform/kvm/machine_arm64_unsafe.go:91:25: possible misuse of unsafe.Pointer
> > /home/travis/.cache/bazel/_bazel_gvisor/bcc3876c18b4288f70c2594c9439fc99/sandbox/processwrapper-sandbox/5/execroot/**main**/pkg/sentry/platform/kvm/machine_arm64_unsafe.go:99:21: possible misuse of unsafe.Pointer
> > Target //runsc:runsc failed to build
> > Use --verbose_failures to see the command lines of failed build steps.

Sorry, this is not blocking. It will be resolved by #2265. (But all that unsafe pointer arithmetic should be fixed and the exception removed eventually.)

Reply:
I think there are two separate issues: 1) data leaking from the ring0.CPU entries, and 2) enforcing alignment of the ring0.CPU struct (so that you can cleanly map it in).

I think it's relatively straight-forward to force an allocation for kvm.vCPU that ensures an alignment of ring0.CPU. As long as kvm.vCPU is a multiple of the system page size (which can be accomplished with some per-arch padding and a sanity test), allocations should always be page-aligned, e.g.: https://golang.org/src/runtime/sizeclasses.go

Reply:
Yes, I only handled the first issue.

And this PR doesn't enforce alignment of the single ring0.CPU nor the newly introduced kernelEntry, just because current code (before this PR) doesn't enforce alignment of the single kvm.vCPU nor ring0.CPU. I want to make the PR as simple as possible, and maybe we can enforce alignment later, or unless you ask me do it now. (The whole CPU entry area has starting and ending alignment to avoid any sensitive data mapping in the upper. But with https://golang.org/src/runtime/sizeclasses.go these alignments might be unneeded, but I think it is OK since it is just 2 pages and it doesn't rely on the undocumented behavior or the runtime)

Enforcing alignment of the ring0.CPU or ring0.kernelEntry struct doesn't mean we can  simply map it in every time when it allocated. We always have to map or prepare the IDT and all the whole percpu "entries" when ring0.Kernel is initialized so that they can be mapped to every APP's cr3. This means mapping the whole kvm.vCPU upper at boot time and changing the allocation/initialization code of the kvm.vCPU which is non-arch dependance.

Besides information leaking, I had tried and found that mapping the whole kvm.vCPU upper was not actually less complex even without enforcing alignment.

Reply:
Okay, I think I have a better understanding of this stuff now.

My ask will be that you just cut down on the complexity here a bit, and make a simple slice of []kernelEntry in the KernelArchState based on the actual maxVCPUs. You can do appropriate alignment at that time, e.g.:

```go
// kernel_unsafe.go
entrySize := unsafe.Sizeof(kernelEntry{})
var (
  entries []kernelEntry
  padding = 1
)
for {
  entries = make([]kernelEntry, maxVCPUs+padding)
  totalSize := (entrySize * (maxVCPUs + padding))
  addr := reflect.ValueOf(&entries[0]).Pointer()
  if addr & (usermem.PageSize-1) == 0 && totalSize >= usermem.PageSize {
      // The runtime forces power-of-2 alignment for allocations, and we are therefore
      // safe once the first address is aligned and the chunk is at least a full page.
      break
  }
  padding = padding << 1
}

Reply:
it is updated as suggested.